consensus/core/src/block.rs

```diff
@@ -195,99 +146,193 @@ impl fmt::Debug for BlockDigest {
 #[allow(unused)]
-#[derive(Deserialize)]
+#[derive(Deserialize, Serialize)]
 pub(crate) struct SignedBlock {
-    block: Block,
+    inner: Block,
     signature: bytes::Bytes,
-
-    #[serde(skip)]
-    serialized: bytes::Bytes,
 }
 
 impl SignedBlock {
-    // TODO: add deserialization and verification.
+    // TODO: add verification.
 }
 
-/// Verifiied block allows access to its content.
+/// VerifiedBlock allows full access to its content.
+/// It should be relatively cheap to copy.
 #[allow(unused)]
-#[derive(Clone, Deserialize, Serialize)]
+#[derive(Clone)]
 pub(crate) struct VerifiedBlock {
-    pub data: Block,
-    pub signature: bytes::Bytes,
+    block: Arc<SignedBlock>,
 
-    #[serde(skip)]
+    // Cached Block digest and serialized SignedBlock, to avoid re-computing these values.
+    digest: BlockDigest,
     serialized: bytes::Bytes,
 }
 
-impl BlockAPI for VerifiedBlock {
-    fn reference(&self) -> BlockRef {
-        self.data.reference()
+impl VerifiedBlock {
+    /// Parses a serialized block from storage, where the block has been verified.
+    /// This should never be called on unverified data received over the network.
+    pub fn from_storage(serialized: bytes::Bytes) -> Result<Self, bcs::Error> {
+        let block: SignedBlock = bcs::from_bytes(serialized.chunk())?;
+        let digest = Self::compute_digest(&block.inner)?;
+        Ok(VerifiedBlock {
+            block: Arc::new(block),
+            digest,
+            serialized,
+        })
     }
 
-    fn digest(&self) -> BlockDigest {
-        self.data.digest()
+    #[cfg(test)]
+    pub(crate) fn new_for_test(block: Block) -> Self {
+        let digest = Self::compute_digest(&block).unwrap();
+        // Use empty signature in test.
+        let signed_block = SignedBlock {
+            inner: block,
+            signature: Default::default(),
+        };
+        let serialized: bytes::Bytes = bcs::to_bytes(&signed_block)
+            .expect("Serialization should not fail")
+            .into();
+        VerifiedBlock {
+            block: Arc::new(signed_block),
+            digest,
+            serialized,
+        }
     }
 
-    fn round(&self) -> Round {
-        self.data.round()
+    /// Returns reference to the block.
+    pub fn reference(&self) -> BlockRef {
+        BlockRef {
+            round: self.round(),
+            author: self.author(),
+            digest: self.digest(),
+        }
     }
 
-    fn author(&self) -> AuthorityIndex {
-        self.data.author()
+    pub fn digest(&self) -> BlockDigest {
+        self.digest
     }
 
-    fn timestamp_ms(&self) -> BlockTimestampMs {
-        self.data.timestamp_ms()
+    /// Returns the serialized block with signature.
+    pub fn serialized(&self) -> &bytes::Bytes {
+        &self.serialized
     }
 
-    fn ancestors(&self) -> &[BlockRef] {
-        self.data.ancestors()
+    fn compute_digest(block: &Block) -> Result<BlockDigest, bcs::Error> {
```

Comment:
In the mysticeti codebase I added `crc32fast` to be used in tests to speed it up. What are your thoughts on doing something similar here? I can add it later when I add the tests.